### ISSUE TITLE
Automation is unreliable

### ISSUE THREAD

**Issue by Kinc4id** [mylar3/mylar3#972]
Sorry I have to post this here, but I tried to register for the forums and for some reason my email address is blocked there. I tried with an alternative email but failed the captcha and now I'm locked out. (Might get an existential crisis now. Am I really human? lol)

I installed mylar3 yesterday on my raspberry pi and it's basically working but the automation is very unreliable.

First it takes hours to snatch issues after adding a series. Is this normal? I'm used to radarr and if you add a movie to radarr it instantly starts a search. I added a series to mylar3 yesterday and the history shows it snatched it 4,5 hours later. Why is this? 

Then not all downloaded issues get post processed. It was a 6 issue series, all 6 issues where snatched and downloaded at the same time but only 3 of them where post processed. I checked the downloaded files and noticed the three issues that didn't get post processed have &#38; in their file name. The original downloaded file didn't have an & in it's filename, mylar3 must have renamed it in a way it can't process it anymore. . Is there an issue with special characters? When I replace &#38; with just a & in the file name and start an manual automated run the file gets imported.

Next thing is that even if mylar3 imports the comics to the library it copies the files to the library folder but doesn't delete the original file. When it imports comics I get this in the log: https://pastebin.com/Dqhbnt1h Strange thing is I did an manual automated run to trigger that error again and for some reason from all the files in the download folder that are also in my library folder it recognized 2 as duplicates and moved them to the duplicate folder. (It left the now empty folder though, is this intended?) So it can move files, but only when moving to the duplicates folder which is a sub folder in the downloads folder. The library folder is on the same drive as the downloads folder. The drive is connected via USB directly to the pi. 

And the last thing is folder monitoring. I have some files in the monitored folder and in the log I can see mylar3 finds them and matches them with the correct comic but doesn't import them because they are not wanted? Do I have to add these comics first to the list in mylar3? I expected mylar3 to add whatever is in this folder. Is there a setting for this?

Edit: And I just now realized that the list on the home screen doesn't update the number of collected issues until you go to the detail page of a comic. 

Edit2: I added around 200 issues today, around 130 were snatched but only 6 of them were imported. In the logs I noticed that in the logs the destination of the download is the temporary folder where nzbget downloads the files to before it moves them to their downloads folder. This temporary folder is on another drive which is also connected per usb to the pi. Could this be the issue? Is mylar3 waiting for the files in the temporary folder but nzbget already moved them to their download folder?

Edit3: I put set the download folder as the nzb get download directory in the mylar3 settings, deleted the file from the hard rive, removed the entry from nzbgets history, set it as wanted in myla3 and searched it again. The file was found and downloaded. While it was moved it was added again to nzbgets queue but skipped because of duplicate. The downloaded file was now post processed and added to my library folder, the original file is still in the download folder. Also I can see a failed post processing attempt in the mylar3 log, I guess this is from the download that failed because it was a duplicate? I tried to copy the log from when I started the search to the newest entry but the pi crashed while I did. This is all a was able to copy: https://pastebin.com/LN7Mm2JF

**Comment by evilhero** [comment]
If you're email/IP is blocked on the forums, it's usually due to a general ban cause of spammers. If you dm me your email address that you're trying to sign up with and your IP (or IP range I guess although that's kinda hit/miss sometimes) - I can unblock it.

Typically this type of convo is for the forums, or especially discord - as github isn't the place to be discussing this as it just convolutes things when we try to look thru issues. That being said, I'll answer this as best I can - but then imma close it with the thought process that you can look into the other avenues of discussion we have for further clarification / assistance.

So first up, the comparison to the *arrs.
When Mylar has to search for items, any item - it has to do it via text based queries. It doesn't have the ability to search indexers by just submitting the imdb-id, or tvdb-id or even musicbrainz-id for precision results. Comics don't get that luxury, and probably never will. What this means is that if you have issue 1 you're searching for, it has to go thru all iterations of the numeric 1 with padding as text parsing doesn't distinguish numbers from text - so it would to need to submit a search for 1, 01 and 001. In addition if the title has punctuation in it, again cause of text-based queries, they have to be submitted in a separate way and then iterate over those numeric issues again. What this means that you could possibly get up to 9/10 api hits alone for one single issue, and with search pauses (which are required) - you're looking at a few minutes usually for it to finish searching if it hasn't found it during the actual search (cause then it will break out and download it). 

Because Mylar is having to submit queries multiple times to the same provider, we can't just send them every second - that goes against most of their TOS regarding API queries (I know no-one reads them, but if you hammer you can lose your access), so they have to spaced out generally at 60s apart (yeah, we play the long game). If we had an id we could query directly against, that would save alot of effort and time - but it's been this way for over a decade and comics will always get the short-end of the stick due to the lack of a data source that's reliable enough to use, and no one really caring enough to change it.

So if you're adding series into Mylar and you have the option **Mark all issues as Wanted** enabled, it will queue up every single issue therein to be searched as above. It doesn't simultaneously search - it queues it up in the order of FIFO (first in, first out). So if you queue up 300 issues with that option enbled, you just told the search queue to queue up 300 issues and go look for them. Then you find another issue you want, and search for that one - that doesn't go to the top of the list, that goes into the 301st spot cause FIFO rules apply (maybe in the future it might allow for some prioritisation of GUI selections over background tasks, but that's a heavy thing to add in). So if you have a large queue already and you add a new series expecting them to get downloaded right away, the rule of FIFO says nope - wait your turn (again, we play the long game). It will get to it eventually, but it has other items that were sent prior to the new ones so they need to be searched for first. Note that there are separate queues for several things - searching, downloading, post-processing, adding - all go by way of the FIFO method of queueing items up. So when a item is found during a search, it gets sent to the download queue. The search queue then goes on to the next item, while the download queue waits for the download to finish. Once it's finished, it hands it off to the post-processing queue to post-process the item. During that time the search found 2 more items and sent them both to the download queue and it's trying to download those....and on and on it goes.

Then you encountered the lovely '&' in the filename - which the indexer that you're using is famous for renaming crap like that from the original files. What this means is that while the response from the indexer to mylar says "this it the file: Series.and.Name.001.2012", Mylar will go ok - I'll make sure to look for that. Only when it's downloaded and the your download client extracts the file the result is 'Series.&.Name.001.2021' which doesn't match up to what Mylar is expecting so it rejects it based on an incorrect match. Both su and dog are the main two that insist on renaming the nzbs so they replace all spaces with decimals, remove all brackets, and sometimes all & with and (or vice-versa, but not usually). This brings with it a whole whack of problems with matching as what happens if there's a decimal in the series, or in the issue number - how's Mylar supposed to know that **Super.1.001.2012** is actually **Super-1 001 (2012)**? It tries, but it just can't do it sometimes.

Ok, so then you start manually post-processing items (not importing - that's a different thing), and things start to get moved around properly and you're thinking this is all shit. The reason why manual post-processing is working is cause it doesn't know what the names are ahead of time, or what the filenames are supposed to be that it's looking for that's being sent by the downloader/provider. It just straight up parses the filenames and goes 'yep, we have that' and spews it out to the correct location.

Then you mention that it doesn't delete the original file during post-processing. Yeah that's cause you have Enforce Permissions turned on within Mylar and it can't set them so it causes the error and can't complete the process of cleaning up the files after they've moved - resulting in files in both locations when there shouldn't be. It's stated in the wiki somewhere (probably buried I know), that Mylar can't enforce permissions outside of the local environment - this would apply to mounted drives that are external to the system in this case (the USB drive). So you need to either fix the permissions on the USB drive where your comics are being moved to that the user running  Mylar has full ownership and access (rwx) to the comic location path and all it's subdirectories. Or, just turn off Enforce Permissions and let your OS handle it, which normally it does via the method that you mount the drive with.

And then finally, like every other automated program out there - Mylar will only post-process (again, not import) issues for series that currently exist on your watchlist. If they do not exist, then it will not be able to post-process them. Every other automation program behaves the same way with this, so it's not like this is any different.  If you have issues that don't belong to a series, and are not on your watchlist (the main/index page is what we call the watchlist), then you need to use the Import A Directory option via the Manage tab. This will import series that don't currently exist on your watchlist, but if a series does exist and you scan it via this option, it will not import it since it already exists (at that point you need to manually post-process it). As a bullet in our FAQ wiki: ``Post-Processing is for adding new issues into existing series on your watchlist, Import is for adding files for series that don't exist on your watchlist into your watchlist``. Mylar doesn't care if the issue is Wanted / Skipped  for the purpose of post-processing items, it will post-process it regardless - unless the status is Downloaded in which case it will check the existing filesize (or filetype based on your file-restrictions type if specified) and either post-process the new item or keep the old item.

The home screen does update the new issues - you can't use the back button on the browser as that will cache items. If it didn't update it's probably because the items weren't finished processing as of yet.

NZBGet also does some weird shit with caching downloads - I know the reasoning, so that it doesn't have to redownload something it already did that  was valid. But when it does that it changes the status to something different, and then the location it's pointing to is like the new-old location within nzbget and it's just confusing af. You have to make sure that any items that fail in nzbget are cleared from the nzbget history - I know it's a pain in the arse, but otherwise the old location is remembered and then what's passed to mylar isn't the correct location so it can't post-process items. Note that this does not occur with SABnzbd - just with NZBGet and how it handles the retry aspect.

If you're not using docker and NZBGet and mylar reside on the same local machine (not drives, like physical machines with different IPs kinda thing), then don't put anything in the download directory location for NZBGet in Mylar. That's forcing mylar to overwrite whatever location is being passed back from NZBGet to what you're indicating which might not be correct in some cases. So yeah, for your use-case, don't use that option.

Ok, so as I stated initially, I'm going to close this as it's not really an issue per-se - but moreso about the workflow and how it all works together. If you want to continue this further, feel free to hop on our discord (the link is in the configuration page of mylar on the 1st tab), and/or send me your email/IP and I can unblock you from the forums (or do both).
